### PR TITLE
More notifications

### PR DESCRIPTION
**Pull request by phil-lopreiato**
Implementation of the rest of our outlined notification types. 

Needs to be thoroughly tested, but I'll take other comments now.

### CODE REVIEW

File: helpers/award_manipulator.py [hunk 1]
```diff
@@ -13,6 +15,20 @@ def getCacheKeysAndControllers(cls, affected_refs):
         return CacheClearer.get_award_cache_keys_and_controllers(affected_refs)
 
     @classmethod
+    def postUpdateHook(cls, awards):
```
Comment (fangeugene) [inline helpers/award_manipulator.py]: This probably needs `updated_attr_list` as an additional arg, even if it's unused.

File: helpers/event_manipulator.py [hunk 2]
```diff
@@ -14,23 +16,33 @@ def getCacheKeysAndControllers(cls, affected_refs):
...
         To run after models have been updated
         """
-        for event in events:
+        for (event, updated_attrs) in zip(events, updated_attr_list):
+            try:
+                if "alliance_selections_json" in updated_attrs:
+                    # Send updated alliances notification
+                    logging.info("Sending alliance notifications for {}".format(event.key_name))
+                    NotificationHelper.send_alliance_update(event)
+                    pass
```
Comment (fangeugene) [inline helpers/event_manipulator.py]: unnecessary

File: helpers/match_manipulator.py [hunk 3]
```diff
@@ -25,13 +26,29 @@ def postUpdateHook(cls, matches):
         Send push notifications to subscribed users
         Only if the match is part of an active event
         '''
+        unplayed_matches = []
```
Comment (fangeugene) [inline helpers/match_manipulator.py]: `updated_attrs` should be an arg to the method.

File: helpers/match_manipulator.py [hunk 4]
```diff
@@ -25,13 +26,29 @@ def postUpdateHook(cls, matches):
...
                     logging.error("Error sending match updates: "+str(exception))
+                    logging.error(traceback.format_exc())
+            elif not match.has_been_played and not event in unplayed_matches:
+                unplayed_matches.append(event)
+
+        
+        '''
+        If we have an unplayed match, send out a schedule update notification
+        '''
+        for event in unplayed_matches:
```
Comment (fangeugene) [inline helpers/match_manipulator.py]: `unplayed_matches` is an odd variable name. `unplayed_match_events` makes more sense. There's probably a better name for it, but I can't think of one at the moment.

File: helpers/match_manipulator.py [hunk 5]
```diff
@@ -25,13 +26,29 @@ def postUpdateHook(cls, matches):
         Send push notifications to subscribed users
         Only if the match is part of an active event
         '''
+        unplayed_matches = []
         for match in matches:
-            if match.event.get().now:
+            event = match.event.get()
+            if event.now and match.has_been_played:
```
Comment (fangeugene) [inline helpers/match_manipulator.py]: The `event.now` test should be outside of both cases.

File: controllers/cron_controller.py [hunk 6]
```diff
@@ -444,3 +445,12 @@ def get(self, event_key):
...
         self.response.out.write(event.district_points)
+
+
+class UpcomingNotificationDo(webapp.RequestHandler):
+    """
+    Sends out notifications for upcoming matches
+    """
+    def get(self):
+        live_events = EventHelper.getEventsWithinADay()
+        NotificationHelper.send_upcoming(live_events)
```
Comment (fangeugene) [inline controllers/cron_controller.py]: `send_upcoming_matches` makes more sense as a method name. Call `EventHelper.getEventsWithinADay()` within that method.

File: helpers/notification_helper.py [hunk 7]
```diff
@@ -40,6 +49,55 @@ def send_subscription_update(cls, user_id, sending_device_key=""):
         notification.send(clients)
 
     @classmethod
+    def send_upcoming(cls, live_events):
+        from helpers.match_helper import MatchHelper
```
Comment (fangeugene) [inline helpers/notification_helper.py]: Any reason why this import is here instead of at the top?

File: helpers/notification_helper.py [hunk 8]
```diff
@@ -40,6 +49,55 @@ def send_subscription_update(cls, user_id, sending_device_key=""):
...
+        now = datetime.datetime.utcnow()
+        for event in live_events:
+            matches = event.matches
+            next_match = MatchHelper.upcomingMatches(matches, num=1)
+            if next_match[0] and not next_match[0].push_sent:
+                match = next_match[0]
+                if match.time is None or (now.day == match.time.day and match.time + datetime.timedelta(minutes=-15) <= now):
```
Comment (fangeugene) [inline helpers/notification_helper.py]: Why are you checking for `now.day == match.time.day` the only events are live events?

File: controllers/cron_controller.py (hunk 6 above)
Comment (phil-lopreiato) [inline controllers/cron_controller.py]: I actually think it's better to pass the events as parameters. I tried to keep all the methods in NotificationHelper dumb as to which models they sent data for. That way, we can send out mock notifications of that type.

File: helpers/notification_helper.py (hunk 7 above)
Comment (phil-lopreiato) [inline helpers/notification_helper.py]: Yes, but I can't remember why. I think it caused an import loop somewhere - I'll test it out after I fix everything else

File: controllers/cron_controller.py (hunk 6 above)
Comment (fangeugene) [inline controllers/cron_controller.py]: Ok fair enough. Do change the method name though.

File: helpers/notification_helper.py (hunk 7 above)
Comment (fangeugene) [inline helpers/notification_helper.py]: Ok comment it if that's the case.

Comment (phil-lopreiato) [inline helpers/notification_helper.py]: Yup, circular import
```
Traceback (most recent call last):
  File "/base/data/home/runtimes/python27/python27_lib/versions/1/google/appengine/runtime/wsgi.py", line 240, in Handle
    handler = _config_handle.add_wsgi_middleware(self._LoadHandler())
  File "/base/data/home/runtimes/python27/python27_lib/versions/1/google/appengine/runtime/wsgi.py", line 299, in _LoadHandler
    handler, path, err = LoadObject(self._handler)
  File "/base/data/home/runtimes/python27/python27_lib/versions/1/google/appengine/runtime/wsgi.py", line 85, in LoadObject
    obj = __import__(path[0])
  File "/base/data/home/apps/s~tba-dev-phil/1-2.382407325469123030/main.py", line 7, in <module>
    from controllers.account_controller import AccountEdit, AccountLogout, AccountOverview, AccountRegister, MyTBAController
  File "/base/data/home/apps/s~tba-dev-phil/1-2.382407325469123030/controllers/account_controller.py", line 16, in <module>
    from helpers.mytba_helper import MyTBAHelper
  File "/base/data/home/apps/s~tba-dev-phil/1-2.382407325469123030/helpers/mytba_helper.py", line 3, in <module>
    from helpers.notification_helper import NotificationHelper
  File "/base/data/home/apps/s~tba-dev-phil/1-2.382407325469123030/helpers/notification_helper.py", line 21, in <module>
    from helpers.match_helper import MatchHelper
  File "/base/data/home/apps/s~tba-dev-phil/1-2.382407325469123030/helpers/match_helper.py", line 8, in <module>
    from helpers.match_manipulator import MatchManipulator
  File "/base/data/home/apps/s~tba-dev-phil/1-2.382407325469123030/helpers/match_manipulator.py", line 10, in <module>
    from helpers.notification_helper import NotificationHelper
ImportError: cannot import name NotificationHelper
```

**Comment by fangeugene** [comment]
This needs to be re merged. I updated the logic at https://github.com/the-blue-alliance/the-blue-alliance/pull/1098/files#diff-5e22a8e2ba5d84b1c091a6224c0b2fccR55 on master

**Comment by phil-lopreiato** [comment]
@fangeugene all set

File: helpers/award_manipulator.py [hunk 9]
```diff
@@ -13,6 +15,23 @@ def getCacheKeysAndControllers(cls, affected_refs):
...
+        # Still needs to be implemented in updateMerge
+        # See helpers.EventManipulator
+        events = []
+        for (award, updated_attrs) in zip(awards, updated_attr_list):
+            event = award.event
+            if event not in events:
+                events.append(event)
+
+        for event in events:
+            try:
```
Comment (fangeugene) [inline helpers/award_manipulator.py]: check if `event.within_a_day`

File: helpers/event_manipulator.py [hunk 10]
```diff
@@ -14,23 +16,32 @@ def getCacheKeysAndControllers(cls, affected_refs):
...
         """
         To run after models have been updated
         """
-        for event in events:
+        for (event, updated_attrs) in zip(events, updated_attr_list):
+            try:
+                if "alliance_selections_json" in updated_attrs:
+                    # Send updated alliances notification
+                    logging.info("Sending alliance notifications for {}".format(event.key_name))
+                    NotificationHelper.send_alliance_update(event)
```
Comment (fangeugene) [inline helpers/event_manipulator.py]: check if `event.within_a_day`

File: helpers/match_manipulator.py [hunk 11]
```diff
@@ -19,19 +20,42 @@ def getCacheKeysAndControllers(cls, affected_refs):
...
-                    NotificationHelper.send_match_score_update(match)
-                except Exception, exception:
-                    logging.error("Error sending match updates: "+str(exception))
+        # Note, updated_attr_list will always be empty, for now
+        # Still needs to be implemented in updateMerge
+        # See helpers.EventManipulator
+        unplayed_match_events = []
+        for (match, updated_attrs) in zip(matches, updated_attr_list):
+            event = match.event.get()
+            if event.now:
```
Comment (fangeugene) [inline helpers/match_manipulator.py]: change to `event.within_a_day`

File: helpers/match_manipulator.py [hunk 12]
```diff
@@ -19,19 +20,42 @@ def getCacheKeysAndControllers(cls, affected_refs):
...
+                    # So send a schedule update notification for the parent event
+                    unplayed_match_events.append(event)
+
+        '''
+        If we have an unplayed match, send out a schedule update notification
+        '''
+        for event in unplayed_match_events:
+            try:
+                logging.info("Sending schedule updates for: {}".format(event.key_name))
+                NotificationHelper.send_schedule_update(event)
```
Comment (fangeugene) [inline helpers/match_manipulator.py]: check if `event.within_a_day`

Comment (phil-lopreiato) [inline helpers/match_manipulator.py]: Good call

**Comment by fangeugene** [comment]
WOOHOO

**Comment by phil-lopreiato** [comment]
#pushAllTheThings

**Pull request merged**